src/Table/extractHeaderLabels.js

```diff
@@ -0,0 +1,71 @@
+import React from 'react'
+
+import { TableHead } from './TableHead'
+import { TableRow } from './TableRow'
+import { TableCellHead } from './TableCellHead'
+
+const extractRowsFromTableChildren = children =>
+    React.Children.toArray(children)
+        // extract theads
+        .filter(child => child.type === TableHead)
+
+        // flatten rows
+        .map(({ props }) => props.children)
+        .reduce((flattened, row) => [...flattened, ...row], [])
+
+        // map row to array of cell components
+        .map(({ props }) => props.children)
+
+const calculateColumnCount = row =>
+    row.reduce(
+        (total, col) =>
+            col.props.colSpan
+                ? total + parseInt(col.props.colSpan, 10)
+                : total + 1,
+        0
+    )
+
+const mapCellsToLabels = row => {
+    let labels = []
+
+    for (let i = 0, count = row.length; i < count; ++i) {
```

Comment:
Maybe I'm missing something obvious, but why are you not comparing `i < row.length` directly here?

Reply:
It's just a habit: https://stackoverflow.com/questions/17989270/for-loop-performance-storing-array-length-in-a-variable

I guess as tables can get ridiculously long with dhis2, this still might be valid, but I don't really have an opinion on this

Reply:
Ah ok. Yeah me neither. It did make me wonder what was going on there, but maybe that's just me.

Reply:
Didn't know that that helped performance.